### PR TITLE
[RIP-15]Add Ipv6 support for RocketMQ

### PR DESCRIPTION
**Pull request by supercym**
## What is the purpose of the change

add the support of IPv6

## Brief changelog

- Modified the message system flag to indicate which IP protocol the client and broker itself came from.
- Modified the method of generating and decoding msgID
- Modified the class related to message storage and decoding

## Verifying this change

https://docs.google.com/presentation/d/1d4GXH0pdwQiUgzlwqJqWLmR5Punz6GG0RvMP0IgHvbM/edit?usp=sharing

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [ ] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [ ] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [ ] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: store/src/test/java/org/apache/rocketmq/store/DefaultMessageStoreTest.java [hunk 1]
```diff
@@ -263,7 +270,7 @@ public void should_return_zero_when_consume_queue_not_found() throws Interrupted
     }
 
     @Test
-    public void should_return_negative_one_when_invoke_getMessageStoreTimeStamp_if_consume_queue_not_found() throws InterruptedException {
+    public void should_return_negative_one_when_invoke_getMessageStoreTimeStamp_if_consume_queue_not_found() throws InterruptedException, UnknownHostException {
```
Comment (vongosling) [inline store/src/test/java/org/apache/rocketmq/store/DefaultMessageStoreTest.java]: Can you modify the test method signature of this section according to code style(test[method]_[conditon] ) in Apache RocketMQ ?

File: store/src/main/java/org/apache/rocketmq/store/CommitLog.java [hunk 2]
```diff
@@ -569,7 +586,19 @@ public PutMessageResult putMessage(final MessageExtBrokerInner msg) {
...
+        if (bornSocketAddress.getAddress() instanceof Inet6Address) {
+            msg.setBornHostV6Flag();
+        }
+
+        InetSocketAddress storeSocketAddress = (InetSocketAddress) msg.getStoreHost();
+        if (storeSocketAddress.getAddress() instanceof Inet6Address) {
+            msg.setStoreHostAddressV6Flag();
+        }
+
+        long eclipseTimeInLock = 0;
```
Comment (liuruiyiyang) [inline store/src/main/java/org/apache/rocketmq/store/CommitLog.java]: The ```eclipseTimeInLock``` should be ```elapsedTimeInLock```?  And there are several places have the same wrong naming, please correct all of them.

File: store/src/main/java/org/apache/rocketmq/store/CommitLog.java [hunk 3]
```diff
@@ -1179,10 +1222,11 @@ public long getJointime() {
 
         private final StringBuilder msgIdBuilder = new StringBuilder();
 
-        private final ByteBuffer hostHolder = ByteBuffer.allocate(8);
+//        private final ByteBuffer hostHolder = ByteBuffer.allocate(8);
```
Comment (liuruiyiyang) [inline store/src/main/java/org/apache/rocketmq/store/CommitLog.java]: If the code is no use, just delete them may be better.

File: store/src/main/java/org/apache/rocketmq/store/CommitLog.java [hunk 4]
```diff
@@ -1432,7 +1499,7 @@ private void resetByteBuffer(final ByteBuffer byteBuffer, final int limit) {
         // The maximum length of the message
         private final int maxMessageSize;
 
-        private final ByteBuffer hostHolder = ByteBuffer.allocate(8);
+//        private final ByteBuffer hostHolder = ByteBuffer.allocate(8);
```
Comment (liuruiyiyang) [inline store/src/main/java/org/apache/rocketmq/store/CommitLog.java]: Just delete the no use code.

File: store/src/test/java/org/apache/rocketmq/store/ConsumeQueueTest.java [hunk 5]
```diff
@@ -127,7 +150,11 @@ protected void putMsg(DefaultMessageStore master) throws Exception {
         long totalMsgs = 200;
 
         for (long i = 0; i < totalMsgs; i++) {
-            master.putMessage(buildMessage());
+            if (i < totalMsgs/2) {
```
Comment (liuruiyiyang) [inline store/src/test/java/org/apache/rocketmq/store/ConsumeQueueTest.java]: Please check the code format even if in the test class.

**Comment by zongtanghu** [comment]
Firstly,I have seen that there are others commits in this pr.So,I suggest you can rebase your pr's codes in you local machine. @supercym

**Comment by zongtanghu** [comment]
I review the pr,initially.I have a important question to ask.
In this pr,you may not take acl ip address pattern matching into account.So I suggest you to see the acl module firstly. @supercym

File: README.md [hunk 6]
```diff
@@ -51,3 +51,23 @@ We always welcome new contributions, whether for trivial cleanups, [big new feat
 [Apache License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0.html) Copyright (C) Apache Software Foundation
```
Comment (zongtanghu) [inline README.md]: Please rebase codes in your local enviroment.

**Comment by ShannonDing** [comment]
@supercym Could you please check the review comments?

File: store/src/test/java/org/apache/rocketmq/store/DefaultMessageStoreTest.java (hunk 1 above)
Comment (supercym) [reply store/src/test/java/org/apache/rocketmq/store/DefaultMessageStoreTest.java]: Thank you for your opinion. It has been modified as required.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -10,7 +10,7 @@ Nor is code the only way to contribute to the project. We strongly value documen
 To submit a change for inclusion, please do the following:
 
 #### If the change is non-trivial please include some unit tests that cover the new functionality.
-#### If you are introducing a completely new feature or API it is a good idea to start a wiki and get consensus on the basic design first.
+#### If you are introducing a completely new feature or API it is a good idea to start a [RIP](https://github.com/apache/rocketmq/wiki/RocketMQ-Improvement-Proposal) and get consensus on the basic design first.
```
[the remark this answers is not in the document]
Comment (supercym) [reply CONTRIBUTING.md]: done！

File: broker/src/main/java/org/apache/rocketmq/broker/processor/PullMessageProcessor.java [hunk 8]
```diff
@@ -494,7 +495,10 @@ public void executeConsumeMessageHookBefore(final ConsumeMessageContext context)
             for (ByteBuffer bb : messageBufferList) {
 
                 byteBuffer.put(bb);
-                storeTimestamp = bb.getLong(MessageDecoder.MESSAGE_STORE_TIMESTAMP_POSTION);
+                int sysFlag = bb.getInt(MessageDecoder.SYSFLAG_POSITION);
+                int bornhostLength = (sysFlag & MessageSysFlag.BORNHOST_V6_FLAG) == 0 ? 8 : 20;
```
[the remark this answers is not in the document]
Comment (supercym) [reply broker/src/main/java/org/apache/rocketmq/broker/processor/PullMessageProcessor.java]: The necessary comments have been added.

File: common/src/main/java/org/apache/rocketmq/common/UtilAll.java [hunk 9]
```diff
@@ -504,6 +540,17 @@ public static String ipToIPv4Str(byte[] ip) {
                                 }
                             }
```
[the remark this answers is not in the document]
Comment (supercym) [reply common/src/main/java/org/apache/rocketmq/common/UtilAll.java]: ok

File: store/src/main/java/org/apache/rocketmq/store/CommitLog.java (hunk 2 above)
Comment (supercym) [reply store/src/main/java/org/apache/rocketmq/store/CommitLog.java]: What you said has been revised.

File: store/src/main/java/org/apache/rocketmq/store/CommitLog.java (hunk 3 above)
Comment (supercym) [reply store/src/main/java/org/apache/rocketmq/store/CommitLog.java]: done!

File: store/src/main/java/org/apache/rocketmq/store/CommitLog.java (hunk 4 above)
Comment (supercym) [reply store/src/main/java/org/apache/rocketmq/store/CommitLog.java]: done!

File: store/src/test/java/org/apache/rocketmq/store/ConsumeQueueTest.java (hunk 5 above)
Comment (supercym) [reply store/src/test/java/org/apache/rocketmq/store/ConsumeQueueTest.java]: Ok, I have finished the work.

File: README.md (hunk 6 above)
Comment (supercym) [reply README.md]: Thank you for your advice. The work is finished.

**Comment by duhenglucky** [comment]
@supercym could you rebase your pr first, this pr brought so many other's commits.

**Pull request merged**